### PR TITLE
Eyedropper improvements

### PR DESCRIPTION
**Pull request by CandyFace**
- Eyedropper will now determine the color from anywhere within the pixel instead of only from the center of the pixel.
- Preview rectangle can now be seen on all backgrounds
- Preview color is now premultiplied, meaning the color represents the inspected pixel rather than the color multiplied with the background.

### CODE REVIEW

File: core_lib/src/tool/eyedroppertool.cpp [hunk 1]
```diff
@@ -143,10 +156,13 @@ void EyedropperTool::updateFrontColor()
...
     BitmapImage* targetImage = layer->getLastBitmapImageAtFrame(mEditor->currentFrame(), 0);
-    if (targetImage == nullptr || !targetImage->contains(getLastPoint())) return QColor();
+    if (targetImage == nullptr || !targetImage->contains(getCurrentPoint())) return QColor();
 
     QColor pickedColour;
-    pickedColour.setRgba(qUnpremultiply(targetImage->pixel(getLastPoint().x(), getLastPoint().y())));
+    const QRgb pixelColor = targetImage->constScanLine(qFloor(qMin(getCurrentPoint().x(),getLastPoint().x())),
+                                                       qFloor(qMin(getCurrentPoint().y(),getLastPoint().y())));
+    pickedColour.setRgba(pixelColor);
+
```
Comment (J5lx) [inline core_lib/src/tool/eyedroppertool.cpp]: As discussed on Discord, the coordinates simply need to be floor()’d here when passed to pixel().

**Review by J5lx — CHANGES_REQUESTED** [review]

File: core_lib/src/tool/eyedroppertool.cpp [hunk 2]
```diff
@@ -63,10 +64,22 @@ QCursor EyedropperTool::cursor(const QColor color)
...
+    int green = static_cast<int>((color.green() * alpha) + (oneMinusAlpha * 255));
+    int blue = static_cast<int>((color.blue() * alpha) + (oneMinusAlpha * 255));
+
     QPainter painter(&pixmap);
     painter.drawPixmap(0, 0, icon);
+    painter.save();
     painter.setPen(QPen(Qt::black, 1, Qt::SolidLine, Qt::RoundCap, Qt::RoundJoin));
-    painter.setBrush(color);
+    painter.restore();
+    painter.drawRect(17, 17, 13, 13);
```
[the remark this answers is not in the document]
Comment (CandyFace) [reply core_lib/src/tool/eyedroppertool.cpp:80]: Yes but not entirely, just enough to create a stroke around the rectangle to cause a multi colored border that will work for all backgrounds.

File: core_lib/src/tool/eyedroppertool.cpp [hunk 3]
```diff
@@ -63,10 +64,22 @@ QCursor EyedropperTool::cursor(const QColor color)
     QPixmap pixmap(32, 32);
     pixmap.fill(Qt::transparent);
 
+
+    /// Color should not blend with background
+    qreal alpha = color.alpha() / 255.0;
+    qreal oneMinusAlpha = 1.0 - alpha;
+    int red = static_cast<int>((color.red() * alpha) + (oneMinusAlpha * 255));
+    int green = static_cast<int>((color.green() * alpha) + (oneMinusAlpha * 255));
+    int blue = static_cast<int>((color.blue() * alpha) + (oneMinusAlpha * 255));
```
[the remark this answers is not in the document]
Comment (CandyFace) [reply core_lib/src/tool/eyedroppertool.cpp]: You're right.. why I thought that was needed. I have simplified it, no need to calculate the blending.

**Review by J5lx — APPROVED** [review]
LGTM!

**Pull request merged**